### ISSUE TITLE
Concept of 'moving a qubit' for ion traps

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#1060]
According to @mlihc15 , their ion traps are divided into three areas: logic area, measurement area, and storage area (in that order in a line). Qubits are moved back and forth between the areas as needed. At the moment they can't re-order the qubits, so this means that if you want to measure qubit number 2 then you need to move qubits 3 through N out of the logic area, past the measurement area, and into the storage area. Then you can move qubit 2 into the measurement area, measure it, and move everything back into the logic area.

This poses a bit of a problem for cirq as-is, because it means the duration of an operation depends on a classical state associated with the qubit (its position in the trap). We don't have that kind of concept.

This issue is a place for suggesting ideas on how to make representing the shuttling around of qubits possible. Personally I see two ways to do it:

1. In addition to gates modifying the quantum state, they can also modify an associated classical state of the circuit. The Device is responsible for tracking this and recomputing it as needed. We have "move qubit" gates which update the classical information.

2. Instead of treating the ions as the qubits, we treat the locations the ions can be as the qubits. So we would have qubits L_0 through L_k for the logic area, M_0 through M_k for the measurement area, and S_0 through S_k for the storage area. We introduce operations for bulk-moving ions between the three regions. Note that it is crucial that simulators understand the bulk-moving operations, since otherwise the simulation cost will cube from 2^k to 2^(3k) since there are three times as many locations as qubits. Alternatively, we could do a rewrite pass on the circuit before passing it into the simulator and this rewrite would relabel qubits instead of moving them.

Personally I'm leaning towards (2) because a) it allows us to better represent the actual scheduling problem and b) having a better concept of "qubit was moved" that allows us to avoid unnecessary overhead in simulations seems good.

**Comment by bryano** [comment]
(2) seems better because the "physical" and "logical" qubit concept isn't restricted to ion traps. In particular, some of the code from the `acquaintance` module can be repurposed.

**Comment by dabacon** [comment]
2 is definitely better.  Certainly there is a sense in which the move operations map down to real hardware commands.  Because I like gates and operations and not new operation classes, I would suggest a fix which is a `ShuffleGate` that corresponds to moving the qubits.  Simulators would need to be updated to support this without doing the permutation directly on the state (i.e. keep a map of qubit locations to qubit locations).  Hardware would, of course, have to do the actual moves.  And the duration depending on locations would be taken care of in `Device` classes.

**Comment by bryano** [comment]
I propose moving `cirq.contrib.acquaintance.permutation` out of contrib. The `PermutationGate` there is what I think @dabacon means by a `ShuffleGate`.

One change I would make is to remove the restriction (which IIRC only affects typing and not runtime) of `LogicalIndex` to `int` and `cirq.Qid`.

**Comment by daxfohl** [comment]
I think #4169 would address the performance implications of (2) here, at least for state vector and density matrix simulators.

**Comment by verult** [comment]
Keeping this before-1.0 in case the final design requires backward-incompatible changes. cc @MichaelBroughton

**Comment by daxfohl** [comment]
I believe this is basically done. When in split_untangled_qubits mode (perhaps we should rename that) you could have 3k qubits, with each position being a qubit in the circuit, but the state vector would never exceed 2^k since only the k qubits would ever be entangled. Also swap is not an entangling operation; internally it just swaps the indices.

To get the final state vector without merging all 3k qubits, there's a `TrialResult.get_state_containing_qubit` you can use. This works with DensityMatrixSimulator and CliffordSimulator too.

**Comment by 95-martin-orion** [comment]
I agree with @daxfohl's assessment - the process described in the top comment can now be expressed using a sequence of SWAPs on the intervening qubits and other operations on the target qubit(s), and (Cirq) simulators will handle it gracefully. Users would need to keep track of which qubits correspond to which locations, but I think that's consistent with the rest of Cirq's design.
